### ISSUE TITLE
jsonconfig generates many compiler warning

### ISSUE THREAD

**Issue by kmaehashi** [jubatus/jubatus_core#132]
When running `./waf --checkall`, I see many `‘value’ may be used uninitialized in this function [-Wuninitialized]` warnings near `jubatus/core/common/jsonconfig/cast.hpp`. Why?

```
In file included from ../jubatus/core/unlearner/../common/jsonconfig.hpp:22:0,
                 from ../jubatus/core/unlearner/unlearner_factory.hpp:21,
                 from ../jubatus/core/unlearner/unlearner_factory.cpp:17:
/home/jubatus/Development/jubatus_core/jubatus/util/text/json/../../data/optional.h: In function ‘void jubatus::core::common::jsonconfig::serialize(jubatus::core::common::jsonconfig::json_config_iarchive_cast&, jubatus::util::data::serialization::named_value<jubatus::util::data::optional<T> >&) [with T = long int]’:
/home/jubatus/Development/jubatus_core/jubatus/util/text/json/../../data/optional.h:83:7: warning: ‘value’ may be used uninitialized in this function [-Wuninitialized]
../jubatus/core/unlearner/../common/./jsonconfig/cast.hpp:247:7: note: ‘value’ was declared here
```

**Comment by gwtnb** [comment]
```value``` contains an uninitialized value, and it is passed by reference, so ```-Wuninitialized``` prints a warning.

**Comment by kmaehashi** [comment]
Thanks! So, how can we eliminate them? (is it possible?)

**Comment by gwtnb** [comment]
We have two solutions:
- replace ```T value;``` with ```T value = T();```
- remove ```T value;``` and pass ```v.v``` instead.

https://github.com/jubatus/jubatus_core/blob/master/jubatus/core/common/jsonconfig/cast.hpp#L240
The latter one breaks strong exception safety, but it seems that strong safety is not required here.

**Comment by rimms** [comment]
Discussion on 2015-03-30:

We decided to solve this problem with the latter.

**Comment by kmaehashi** [comment]
> * remove T value; and pass v.v instead.

It seems that this idea does not work, as `jubatus::util::data::optional` does not implement `serialize` method (so that implicit type conversion does not work)?

**Comment by kmaehashi** [comment]
Fixed via #141
